File: pkg/landgrab/landgrab.go
package landgrab

import (
	"fmt"
	"math/rand"
	"sort"

	"github.com/montanaflynn/stats"
)

const diceSides = 6

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func roll(numDice int) []int {
	ret := make([]int, numDice)
	for i := 0; i < numDice; i++ {
		ret[i] = rand.Intn(diceSides) + 1
	}
	sort.Sort(sort.Reverse(sort.IntSlice(ret)))
	return ret
}

func oneRound(attackers, defenders, dice int) (int, int) {
	if dice > attackers {
		dice = attackers
	}
	attackerDice := min(dice, attackers)
	defenderDice := min(2, defenders)

	matchups := min(attackerDice, defenderDice)

	attackerRoll := roll(attackerDice)
	defenderRoll := roll(defenderDice)

	for i := 0; i < matchups; i++ {
		if attackerRoll[i] > defenderRoll[i] {
			defenders--
		} else {
			attackers--
		}
	}
	return attackers, defenders
}

// Attack with as much as you ahve until it's either taken or you can't attack
func invade(attackers, defenders int) (int, int, bool) {

	for attackers > 1 && defenders > 0 {
		attackers, defenders = oneRound(attackers, defenders, 3)
	}
	if defenders == 0 {
		return 1, attackers - 1, true
	}
	return attackers, defenders, false
}

// Given an attacking territory w attackers, invade one territory after another
// Return how many attacking armies are in the last territory, or 0 if you never get there
func campaign(attackers int, defenderingTerritories []int) int {
	for i := 0; i < len(defenderingTerritories); i++ {
		_, newAttackers, success := invade(attackers, defenderingTerritories[i])
		//fmt.Printf("Invading %d->%d leaves %d in defending w success %v\n", attackers, defenderingTerritories[i], newAttackers, success)
		attackers = newAttackers
		if !success {
			return 0
		}
	}
	return attackers
}

func probabilityDesiredRemaining(attackers int, defendingTerritories []int, trials int) (float64, float64, float64, float64) {
	successes := 0.0
	const batches = 1

	results := make([]int, trials)

	for i := 0; i < trials; i++ {
		//fmt.Println(i)
		remaining := campaign(attackers, defendingTerritories)
		if remaining != 0 {
			successes += 1
		}

		results[i] = remaining
		//	fmt.Println(i)
	}
	resultsForStats := stats.LoadRawData(results)
	p10, _ := resultsForStats.Percentile(10)
	p50, _ := resultsForStats.Percentile(50)
	p90, _ := resultsForStats.Percentile(90)
	floatTrials := float64(trials)
	return p10, p50, p90, (successes / floatTrials) * 100
}

type attackerSummary struct {
	attackers int
	p10       float64
	p50       float64
	p90       float64
	prob      float64
}

func (a attackerSummary) String() string {
	probStr := "100  "
	if a.prob < 99.999 {
		probStr = fmt.Sprintf("%05.2f", a.prob)
	}

	return fmt.Sprintf("%-7d%s    %-3d   %-3d   %-3d", a.attackers, probStr, int(a.p10), int(a.p50), int(a.p90))
}

func getAttackerSummary(attackers int, defendingTerritories []int, trialsPerAttacker int) attackerSummary {
	p10, p50, p90, prob := probabilityDesiredRemaining(attackers, defendingTerritories, trialsPerAttacker)
	return attackerSummary{
		attackers: attackers,
		p10:       p10,
		p50:       p50,
		p90:       p90,
		prob:      prob,
	}
}

func worker(defendingTerritories []int, trialsPerAttacker int, results chan attackerSummary, attackersChannel chan int) {
	for {
		attackers := <-attackersChannel
		results <- getAttackerSummary(attackers, defendingTerritories, trialsPerAttacker)
	}
}

func DetermineAttackers(defendingTerritories []int) {
	//attackers := 1
	trialsPerAttacker := 10_000
	threads := 1
	totalDefendingArmies := 0
	for i := 0; i < len(defendingTerritories); i++ {
		totalDefendingArmies += defendingTerritories[i]
	}
	fmt.Printf("Calculating size of force needed to defeat %d armies to claim %d territories\n", totalDefendingArmies, len(defendingTerritories))
	fmt.Println("Attack Success  p10   p50   p90")
	results := make(chan attackerSummary, 100)
	attackersChannel := make(chan int, 10)
	go func() {
		for i := 0; ; i++ {
			//fmt.Printf("Please work on %d\n", i)
			attackersChannel <- i
		}
	}()
	for i := 0; i < threads; i++ {
		go worker(defendingTerritories, trialsPerAttacker, results, attackersChannel)
	}
	for {
		summary := <-results
		fmt.Printf("%s\n", summary)
		if summary.prob > 99.99 {
			return
		}

	}
}

File: landgrab.py
#!/usr/bin/env python3

import random
import argparse

DICE_SIDES = 6

def roll(num_dice):
    ret = []
    for i in range(num_dice):
        ret.append(random.randint(0, DICE_SIDES) + 1)
    return sorted(ret, reverse=True)



def one_round(attackers, defenders, dice):
    """
    One round of attacks, returning new army levels
    """
    if dice > attackers:
        dice = attackers
    attacker_dice = min(dice, attackers)
    defender_dice = min(2, defenders)

    matchups = min(attacker_dice, defender_dice)

    attacker_role = roll(attacker_dice)
    defender_role = roll(defender_dice)

    for i in range(matchups):
        # Attacker's role must be larger than defenders,
        # tie goes to defender
        if attacker_role[i] > defender_role[i]:
            defenders-=1
        else:
            attackers-=1
    return attackers, defenders


def invade(attackers, defenders):
    """
    Attack with as much as you have until it's either taken over or you can't attack
    """
    while attackers >1 and defenders > 0:
        attackers, defenders = one_round(attackers, defenders, 3)
    # Decision here to march all troops in after winning.
    # If we decide to only march some, you have to keep track of
    # how many attacking dice were rolled, because at least that many
    # have to march
    if defenders == 0:
        # Defender lost, move all troops
        return 1, attackers-1, True

    return attackers, defenders, False

def campaign(attackers, defending_territories):
    """
    Given an attacking territory w attackers, invade one territory after another
    Return how many attacking armies are in the last territory, or 0 if you never get there
    """
    for defending_territory in defending_territories:
        # After invasion, we will leave behind everyone, and whoever is in the previously
        # defending territory will become next attackers. If invasion failed, we quit now
        _, attackers, success = invade(attackers, defending_territory)
        #print(f"Invading {attackers}->{defending_territory} leaves {newAttackers} in defending w success {success}")
        if not success:
            return 0
    return attackers


def probability_desired_remaining(attackers, defending_territories, desired_remaining_attackers, trials):
    # TODO: Make the code figure out num of correct trials based on some tolerance
    success = 0.0
    total = 0.0
    failures = 0.0
    for i in range(trials):
        remaining = campaign(attackers, defending_territories)
        #print(f"attacking with {attackers} against {defending_territories} left {remaining}")
        if remaining >= desired_remaining_attackers:
            success+=1.0
        if remaining == 0:
            failures+=1
        total += remaining
    return success/trials * 100, total/trials, failures/trials * 100



def determine_attackers(desired_remaining_attackers, defending_territories):
    attackers = 1
    trials_per_attacker = 10_000
    print("Attack Success  Failure Avg")
    while True:
        prob, avg, prob_failure = probability_desired_remaining(attackers, defending_territories, desired_remaining_attackers, trials_per_attacker)
        if prob_failure > 99.999:
            prob_failure_str="100  "
        else:
            prob_failure_str="%05.2f" % (prob_failure)
        attackers+=1
        print("%-7s%05.2f    %s  %2s" % (attackers, prob, prob_failure_str, int(avg)))
        if prob > 99:
            break


def main():
    print("DEPRECATED! Use the go version")
    return

    parser = argparse.ArgumentParser()
    parser.add_argument("remaining",metavar="A", type=int)
    parser.add_argument("territories", metavar="N", type=int, nargs="+")

    args = parser.parse_args()
    print(f"Calculating number of attackers needed to attack {args.territories} and still have {args.remaining} remaining ... ")
    
    determine_attackers(args.remaining, args.territories)


if __name__ == "__main__":
    main()
